[FSDP] Implement apply()

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #72600

Implements `apply()` which applies a `callable` of signature `f(m: Module) -> None` recursively to every submodule. The main difference from `nn.module.apply` is that this version summons the full parameters before apply() so it works appropriately with FSDP.

Closes https://github.com/pytorch/pytorch/issues/72554

Differential Revision: [D34111109](https://our.internmc.facebook.com/intern/diff/D34111109/)

rohan-varma committed:
[FSDP] Implement apply()

Implements `apply()` which applies a `callable` of signature `f(m: Module) -> None` recursively to every submodule. The main difference from `nn.module.apply` is that this version summons the full parameters before apply() so it works appropriately with FSDP.

Differential Revision: [D34111109](https://our.internmc.facebook.com/intern/diff/D34111109/)

[ghstack-poisoned]

zhaojuanmao commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -287,6 +287,45 @@ def module(self) -> FlattenParamsWrapper:
+        """
+        uninitialized = self._is_root is None
+        self._assert_state(TrainingState_.IDLE)
+        with self._summon_full_params(recurse=False):
```
hmm, the comment said 'Applies ``fn`` recursively to every submodule', but here _summon_full_params is called with recurse=False, not sure how it works

rohan-varma replied:
It is similar to how `state_dict` will be implemented, we don't recurse as we call `super().apply()` which calls the method in nn.Module. That method should run `fn` on this module and then recursively call in on its children modules, so as a result the FSDP instances will have `apply` invoked so it makes sense for it to be nonrecursive.

zhaojuanmao replied:
I see, that makes sense! Thanks for explaining

zhaojuanmao reviewed (approved):
looks good, just having one question

rohan-varma committed:
Update on "[FSDP] Implement apply()"


Implements `apply()` which applies a `callable` of signature `f(m: Module) -> None` recursively to every submodule. The main difference from `nn.module.apply` is that this version summons the full parameters before apply() so it works appropriately with FSDP.

Closes https://github.com/pytorch/pytorch/issues/72554

Differential Revision: [D34111109](https://our.internmc.facebook.com/intern/diff/D34111109/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[FSDP] Implement apply()"


Implements `apply()` which applies a `callable` of signature `f(m: Module) -> None` recursively to every submodule. The main difference from `nn.module.apply` is that this version summons the full parameters before apply() so it works appropriately with FSDP.

Closes https://github.com/pytorch/pytorch/issues/72554

Differential Revision: [D34111109](https://our.internmc.facebook.com/intern/diff/D34111109/)

[ghstack-poisoned]
